File: src/main/java/com/bajdcc/LALR1/grammar/tree/ExpBinop.java
package com.bajdcc.LALR1.grammar.tree;

import com.bajdcc.LALR1.grammar.codegen.ICodegen;
import com.bajdcc.LALR1.grammar.error.SemanticException;
import com.bajdcc.LALR1.grammar.runtime.RuntimeInst;
import com.bajdcc.LALR1.grammar.runtime.RuntimeInstUnary;
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder;
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope;
import com.bajdcc.LALR1.grammar.type.TokenTools;
import com.bajdcc.util.lexer.token.OperatorType;
import com.bajdcc.util.lexer.token.Token;
import com.bajdcc.util.lexer.token.TokenType;

/**
 * 【语义分析】二元表达式
 *
 * @author bajdcc
 */
public class ExpBinop implements IExp {

	/**
	 * 操作符
	 */
	private Token token = null;

	/**
	 * 左操作数
	 */
	private IExp leftOperand = null;

	/**
	 * 右操作数
	 */
	private IExp rightOperand = null;

	public Token getToken() {
		return token;
	}

	public void setToken(Token token) {
		this.token = token;
	}

	public IExp getLeftOperand() {
		return leftOperand;
	}

	public void setLeftOperand(IExp leftOperand) {
		this.leftOperand = leftOperand;
	}

	public IExp getRightOperand() {
		return rightOperand;
	}

	public void setRightOperand(IExp rightOperand) {
		this.rightOperand = rightOperand;
	}

	@Override
	public boolean isConstant() {
		return leftOperand.isConstant() && rightOperand.isConstant();
	}

	@Override
	public boolean isEnumerable() {
		return leftOperand.isEnumerable() ^ rightOperand.isEnumerable();
	}

	@Override
	public IExp simplify(ISemanticRecorder recorder) {
		if (!isConstant()) {
			return this;
		}
		if (leftOperand instanceof ExpValue && rightOperand instanceof ExpValue) {
			if (token.kToken == TokenType.OPERATOR) {
				OperatorType op = (OperatorType) token.object;
				if (op == OperatorType.COLON)
					return this;
				if (TokenTools.binop(recorder, this)) {
					return leftOperand;
				}
			}
		}
		return this;
	}

	@Override
	public void analysis(ISemanticRecorder recorder) {
		if (token.kToken == TokenType.OPERATOR) {
			OperatorType op = (OperatorType) token.object;
			if (TokenTools.isAssignment(op)) {
				if (!(leftOperand instanceof ExpValue)) {
					recorder.add(SemanticException.SemanticError.INVALID_ASSIGNMENT, token);
				}
			}
		}
		leftOperand.analysis(recorder);
		rightOperand.analysis(recorder);
	}

	@Override
	public void genCode(ICodegen codegen) {
		if (token.kToken == TokenType.OPERATOR && token.object == OperatorType.DOT) {
			codegen.genCode(RuntimeInst.iopena);
			leftOperand.genCode(codegen);
			codegen.genCode(RuntimeInst.ipusha);
			rightOperand.genCode(codegen);
			codegen.genCode(RuntimeInst.ipusha);
			codegen.genCode(RuntimeInst.ipush, codegen.genDataRef("g_get_property"));
			codegen.genCode(RuntimeInst.icallx);
			return;
		}
		if (token.kToken == TokenType.OPERATOR) {
			OperatorType op = (OperatorType) token.object;
			if (TokenTools.isAssignment(op)) {
				RuntimeInst ins = TokenTools.op2ins(token);
				ExpValue left = (ExpValue) leftOperand;
				if (ins == RuntimeInst.ice) {
					rightOperand.genCode(codegen);
					codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(left.getToken().object));
					codegen.genCode(RuntimeInst.istore);
					return;
				}
				leftOperand.genCode(codegen);
				rightOperand.genCode(codegen);
				codegen.genCode(ins);
				codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(left.getToken().object));
				codegen.genCode(RuntimeInst.istore);
				return;
			} else if (op == OperatorType.COLON) {
				rightOperand.genCode(codegen);
				leftOperand.genCode(codegen);
				return;
			}
		}
		RuntimeInst inst = TokenTools.op2ins(token);
		leftOperand.genCode(codegen);
		RuntimeInstUnary jmp = null;
		switch (inst) {
			case iandl:
				jmp = codegen.genCode(RuntimeInst.ijfx, -1);
				break;
			case iorl:
				jmp = codegen.genCode(RuntimeInst.ijtx, -1);
				break;
			default:
				break;
		}
		rightOperand.genCode(codegen);
		codegen.genCode(inst);
		if (jmp != null) {
			jmp.op1 = codegen.getCodeIndex();
		}
	}

	@Override
	public String toString() {
		return print(new StringBuilder());
	}

	@Override
	public String print(StringBuilder prefix) {
		if (token.kToken == TokenType.OPERATOR && token.object == OperatorType.COLON) {
			return leftOperand.print(prefix) + " " + token.toRealString() + " "
					+ rightOperand.print(prefix);
		}
		return "( " + leftOperand.print(prefix) + " " + token.toRealString() + " "
				+ rightOperand.print(prefix) + " )";
	}

	@Override
	public void addClosure(IClosureScope scope) {
		leftOperand.addClosure(scope);
		rightOperand.addClosure(scope);
	}

	@Override
	public void setYield() {
		leftOperand.setYield();
		rightOperand.setYield();
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/tree/ExpBinop.kt
package com.bajdcc.LALR1.grammar.tree

import com.bajdcc.LALR1.grammar.codegen.ICodegen
import com.bajdcc.LALR1.grammar.error.SemanticException
import com.bajdcc.LALR1.grammar.runtime.RuntimeInst
import com.bajdcc.LALR1.grammar.runtime.RuntimeInstUnary
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope
import com.bajdcc.LALR1.grammar.type.TokenTools
import com.bajdcc.util.lexer.token.OperatorType
import com.bajdcc.util.lexer.token.Token
import com.bajdcc.util.lexer.token.TokenType

/**
 * 【语义分析】二元表达式
 *
 * @author bajdcc
 */
class ExpBinop : IExp {

    /**
     * 操作符
     */
    var token: Token? = null

    /**
     * 左操作数
     */
    var leftOperand: IExp? = null

    /**
     * 右操作数
     */
    var rightOperand: IExp? = null

    override fun isConstant(): Boolean {
        return leftOperand!!.isConstant() && rightOperand!!.isConstant()
    }

    override fun isEnumerable(): Boolean {
        return leftOperand!!.isEnumerable() xor rightOperand!!.isEnumerable()
    }

    override fun simplify(recorder: ISemanticRecorder): IExp {
        if (!isConstant()) {
            return this
        }
        if (leftOperand is ExpValue && rightOperand is ExpValue) {
            if (token!!.type === TokenType.OPERATOR) {
                val op = token!!.obj as OperatorType?
                if (op === OperatorType.COLON)
                    return this
                if (TokenTools.binop(recorder, this)) {
                    return leftOperand!!
                }
            }
        }
        return this
    }

    override fun analysis(recorder: ISemanticRecorder) {
        if (token!!.type === TokenType.OPERATOR) {
            val op = token!!.obj as OperatorType?
            if (TokenTools.isAssignment(op!!)) {
                if (leftOperand !is ExpValue) {
                    recorder.add(SemanticException.SemanticError.INVALID_ASSIGNMENT, token)
                }
            }
        }
        leftOperand!!.analysis(recorder)
        rightOperand!!.analysis(recorder)
    }

    override fun genCode(codegen: ICodegen) {
        if (token!!.type === TokenType.OPERATOR && token!!.obj === OperatorType.DOT) {
            codegen.genCode(RuntimeInst.iopena)
            leftOperand!!.genCode(codegen)
            codegen.genCode(RuntimeInst.ipusha)
            rightOperand!!.genCode(codegen)
            codegen.genCode(RuntimeInst.ipusha)
            codegen.genCode(RuntimeInst.ipush, codegen.genDataRef("g_get_property"))
            codegen.genCode(RuntimeInst.icallx)
            return
        }
        if (token!!.type === TokenType.OPERATOR) {
            val op = token!!.obj as OperatorType?
            if (TokenTools.isAssignment(op!!)) {
                val ins = TokenTools.op2ins(token!!)
                val left = leftOperand as ExpValue?
                if (ins == RuntimeInst.ice) {
                    rightOperand!!.genCode(codegen)
                    codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(left!!.token!!.obj!!))
                    codegen.genCode(RuntimeInst.istore)
                    return
                }
                leftOperand!!.genCode(codegen)
                rightOperand!!.genCode(codegen)
                codegen.genCode(ins)
                codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(left!!.token!!.obj))
                codegen.genCode(RuntimeInst.istore)
                return
            } else if (op === OperatorType.COLON) {
                rightOperand!!.genCode(codegen)
                leftOperand!!.genCode(codegen)
                return
            }
        }
        val inst = TokenTools.op2ins(token!!)
        leftOperand!!.genCode(codegen)
        var jmp: RuntimeInstUnary? = null
        when (inst) {
            RuntimeInst.iandl -> jmp = codegen.genCode(RuntimeInst.ijfx, -1)
            RuntimeInst.iorl -> jmp = codegen.genCode(RuntimeInst.ijtx, -1)
            else -> {
            }
        }
        rightOperand!!.genCode(codegen)
        codegen.genCode(inst)
        if (jmp != null) {
            jmp.op1 = codegen.codeIndex
        }
    }

    override fun toString(): String {
        return print(StringBuilder())
    }

    override fun print(prefix: StringBuilder): String {
        return if (token!!.type === TokenType.OPERATOR && token!!.obj === OperatorType.COLON) {
            (leftOperand!!.print(prefix) + " " + token!!.toRealString() + " "
                    + rightOperand!!.print(prefix))
        } else "( " + leftOperand!!.print(prefix) + " " + token!!.toRealString() + " " + rightOperand!!.print(prefix) + " )"
    }

    override fun addClosure(scope: IClosureScope) {
        leftOperand!!.addClosure(scope)
        rightOperand!!.addClosure(scope)
    }

    override fun setYield() {
        leftOperand!!.setYield()
        rightOperand!!.setYield()
    }
}
